cherab/tools/observers/bolometry.py

```diff
@@ -276,229 +227,151 @@ def __init__(self, detector_id, centre_point, basis_x, dx, basis_y, dy, slit,
-
-        world = self.root
-
-        self._los_radiance_sensitivity = SensitivityMatrix(grid, self.detector_id, 'los radiance')
-        self._volume_radiance_sensitivity = SensitivityMatrix(grid, self.detector_id, 'Volume mean radiance sensitivity')
-        self._volume_power_sensitivity = SensitivityMatrix(grid, self.detector_id, 'Volume power sensitivity')
-
-        # Make a uniform emitter with 1 W/str/m^3
-        wvl_range = self._volume_observer.max_wavelength - self._volume_observer.min_wavelength
-        emitter = UniformVolumeEmitter(ConstantSF(1/wvl_range))
-
-        if cell_range is None:
-            cell_range = range(grid.count)
-
-        for i in cell_range:
-
-            p1, p2, p3, p4 = grid[i]
-
-            r_inner = p1.x
-            r_outer = p3.x
-            if r_inner > r_outer:
-                r_inner, r_outer = r_outer, r_inner
-
-            z_lower = p2.y
-            z_upper = p1.y
-            if z_lower > z_upper:
-                z_lower, z_upper = z_upper, z_lower
-
-            # TODO - switch to using CAD method such that reflections can be included automatically
-            cylinder_height = z_upper - z_lower
-
-            outer_cylinder = Cylinder(radius=r_outer, height=cylinder_height, transform=translate(0, 0, z_lower))
-            inner_cylinder = Cylinder(radius=r_inner, height=cylinder_height, transform=translate(0, 0, z_lower))
-            cell_emitter = Subtract(outer_cylinder, inner_cylinder, parent=world, material=emitter)
-
-            self._los_observer.observe()
-            self._los_radiance_sensitivity.sensitivity[i] = self._los_radiance_pipeline.value.mean
-
-            self._volume_observer.observe()
-            self._volume_radiance_sensitivity.sensitivity[i] = self._volume_radiance_pipeline.value.mean
-            self._volume_power_sensitivity.sensitivity[i] = self._volume_power_pipeline.value.mean
-
-            outer_cylinder.parent = None
-            inner_cylinder.parent = None
-            cell_emitter.parent = None
-
-    def calculate_etendue(self, ray_count=10000, batches=10, print_results=False):
-
-        if self.slit.csg_aperture is None:
-            raise ValueError("CSG aperture is required to support etendue calculation.")
+    def trace_sightline(self):
+
+        if not isinstance(self.root, World):
+            raise ValueError("This BolometerFoil is not connected to a valid World scenegraph object.")
+
+        centre_point = self.centre_point
+
+        while True:
+
+            # Find the next intersection point of the ray with the world
+            intersection = self.root.hit(CoreRay(centre_point, self.sightline_vector))
+
+            if intersection is None:
+                raise RuntimeError("No material intersection was found for this sightline.")
+
+            elif isinstance(intersection.primitive.material, NullMaterial):
+                centre_point += self.sightline_vector * 1E-4
```

Comment:
Nice magic number. Any reason why 1E-4? Can we use the intersection primitive's width to choose an appropriate point to start the ray again instead?

Reply:
Agreed, I've refactored the code to calculate a displacement based on the minimum(dx, dy) / 100.